README.md
# JS-Lib

## Http

App Masters fetch class.

### Configuration

On the project setup, set the base url for your API and the necessary headers:

````javascript
import { Http } from '@app-masters/js-lib';

...

const version = '1.0.0'; // get from package.json
const client = 'mobile'; // use 'mobile', 'admin' or 'web'
const env = 'development'; // use 'development', 'staging' or 'production'
const contentType = 'application/json';
const baseURL = 'http://localhost.com:3000';
const token = 'JWT-0000000000000'; // user's authentication token

Http.setup(version, client, env, contentType);
Http.setBaseURL(baseURL);

// after the authentication process
Http.setAuthorization(token);

````

Setting your headers like this, will give you this headers:

````javascript
{
    'Accept': 'application/json',
    'Content-Type': 'application/json',
    'mobile-version': '1.0.0',
    'client': 'mobile',
    'client-env': 'development',
    'authorization': 'JWT-0000000000000'
}
````

Also, you can set a endpoint param. This param is a key/value that will overwrite the key with the value every time that it's found in a fetch call.

Examples:

````javascript

Http.setEndpointParam('{_userID}', user.id);
Http.setEndpointParam('{_maxItems}', 5);

...

Http.get('/{_userID)');
Http.get('/items/?quantity={_maxItems}');
````

Anytime you can reset the headers to the default setup using ``Http.reset()``. This will reset your headers to what was set in ``Http.setup()``, removing the ``authorization`` and all endpoint parameters.

### Usage

The Http class have 4 main asynchronous methods for fetch: ``GET``, ``POST``, ``PUT`` and ``DELETE``. After you set the base URL, you only need to pass the endpoint to fetch it.

Examples:

````javascript

Http.get('/users').then(response => {
    console.log(response);
}).catch(error => {
    console.error(error);
});

// Using endpoint params
Http.get('/users/{_userID}').then(response => {
    console.log(response);
}).catch(error => {
    console.error(error);
});

Http.post('/users', {email: 'new@email.com'}).then(response => {
    console.log(response);
}).catch(error => {
    console.error(error);
});

// Full URLs ignore the baseURL set.
Http.post('http://localhost.com:3001/otherUsers', {email: 'new@email.com'}).then(response => {
    console.log(response);
}).catch(error => {
    console.error(error);
});

Http.put('/users/TH1S154U53R1D', {email: 'other@email.com'}).then(response => {
    console.log(response);
}).catch(error => {
    console.error(error);
});

Http.delete('/users/TH1S154U53R1D').then(response => {
    console.log(response);
}).catch(error => {
    console.error(error);
});
````

The status codes considered as success are: 200, 201 and 204 (on delete).

## AppBoostrap

App Masters default app setup process.

Create a config file with defaults environments:

```

const envs = {};

envs.development = {
    baseUrl: "http://localhost:3000/api",
    uploadCareToken: "someXPTOtoken",
    rollbarToken: "anotherCrazyToken"
};

envs.development_firebase = {
    baseUrl: "https://remotehost-dev.herokuapp.com:3000/api",
    uploadCareToken: "someXPTOtoken",
    rollbarToken: "anotherCrazyToken"
};

envs.staging = {
    baseUrl: "https://remotehost-staging.herokuapp.com:3000/api",
    uploadCareToken: "someXPTOtoken",
    rollbarToken: "anotherCrazyToken"
};

envs.production = {
    baseUrl: "https://remotehost-prod.herokuapp.com:3000/api",
    uploadCareToken: "someXPTOtoken",
    rollbarToken: "anotherCrazyToken"
};

export default envs;

```

Import and call AppBootstrap:


````
import {store, storage} from './store';
import packag from '../package.json';
import envs from './config';
import {AppBootstrap} from '@app-masters/js-lib';

class App extends Component {
    constructor () {
        super();
        try {
            let callbacks = {
                onMinVersionNotSatifies: (version) => {
                    alert("Você deve atualizar sua versão agora! Por favor recarregue a página, se a mensagem continuar, limpe o cache do navegador.");
                },
                onNewVersion: (version) => {
                    alert('Bem vindo à nova versão!');
                    if (version === '0.2.0') {
                      // Some new version decision
                    }
                },
                onUncaughtError: (e) => {
                    if (e.message !== 'Failed to fetch') {
                        Rollbar.error(e);
                        alert("Houve um erro inesperado e os programadores responsáveis já foram avisados. \n\n Detalhes técnicos: " + e.message);
                    } else {
                        alert('Falha na conexāo');
                    }
                }
            };

            // Bootstrap
            AppBootstrap.setup("web", packag, envs, storage, callbacks);
        } catch (e) {
            alert("Houve um erro inesperado e os programadores responsáveis já foram avisados. \n\n Detalhes técnicos: " + e.message);
        }
    }
}

````

(callbacks are the same object used with AMActions.setup()).

## Notification (web)
### Service worker configuration
- Create a file called 'firebase-messaging-sw.js' with the following content and put it on the public path of the project (generaly it is project-folder/public):
```javascript
importScripts('https://www.gstatic.com/firebasejs/4.8.1/firebase-app.js');
importScripts('https://www.gstatic.com/firebasejs/4.8.1/firebase-messaging.js');
firebase.initializeApp({
    messagingSenderId: "1072711350296"
});
const Notifier = firebase.messaging();
Notifier.setBackgroundMessageHandler(function(payload) {
    console.log('[firebase-messaging-sw.js] Received background message ', payload);
    // Customize notification here
    const notificationTitle = 'Background Message Title';
    const notificationOptions = {
        body: 'Background Message body.',
        icon: '/firebase-logo.png'
    };
    return self.registration.showNotification(notificationTitle, notificationOptions);
});
```
- Add the following line on the index.html before the closing tag: `</html>`
```html
<html>
    <head>...</head>
    <body>...</body>
    <!-- add this script -->
    <script src="https://www.gstatic.com/firebasejs/4.4.0/firebase.js"></script>
</html>
```

### Firebase initialize and lib setup
-  Initialize the firebase and then setup the notification lib
```javascript
// initializing the firebase
firebase.initializeApp({
    apiKey: '',
    authDomain: '',
    databaseURL: '',
    projectId: '',
    storageBucket: '',
    messagingSenderId: ''
});
// lib setup
Notification.setup(firebase);
Notification.getToken((token) => console.log('It has the token! Notification should work fine', token));
Notification.handleMessage((message) => {console.log('message received -> ', message)}, console.error);
```

### Lib methods
#### setup
- Receives the config param (can be a firebase instance or config params to set the cordova firebase plugin for example), sets the device info (cordova, web, native) and finnaly sets the messaging (object used to make all the firebase calls).

# Change Log

Check all changes on [changelog](CHANGELOG.md).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
setHeaderParam

## [1.2.6/1.2.7] - 2019-01-11
### Added
- customEnv on appBoostrap
### Changed
- customEnv log

## [1.2.0/1.2.5] - 2019-01-03
### Added
- LaravelErrorHandler for deal with the new error standards
### Changed
- LaravelErrorHandler handling string and object payload
- Accepting larger range of status
- Removing string message from error
- Removing useless fields from error
- Setting 'user' error level

## [1.1.55] - 2018-11-07
### Changed
- Added humanizeDistance on Text

## [1.1.51] - 2018-11-07
### Changed
- Changed [DateTime.relative](/src/dateTime.js) to use `moment`

## [1.1.51] - 2018-11-07
### Changed
- Changed Plural into a singleton

## [1.1.50] - 2018-11-07
### Changed
- Added Plural module

## [1.1.44] - 2018-11-07
### Changed
- Removing body from http.delete

# [1.1.41] - 2018-08-21
### Fixed
- Changed compiling to prepublishOnly and changed to Babel 7

# [1.1.32] - 2018-08-21
### Fixed
- Fix on DateTime.relative

# [1.1.31] - 2018-08-21
### Added
- Text.isNormalInteger added, checks if a string is a number

## [1.1.28/1.1.30] - 2018-08-02
### Changed
- Text.slugify accept a replace object param 

## [1.1.27] - 2018-07-13
### Changed
- Fix to call ucWords on firstname


## [1.1.23] - 2018-05-04
### Added
- setHeaderParam method on HTTP

## [1.1.24] - 2018-05-04
- added "same origin" param on HTTP

## [1.1.22] - 2018-03-09
### Fixed
- Fixing checkToken and setToken

## [1.1.21] - 2018-03-05
### Changed
- Checking the permission before getting the token

## [1.1.20] - 2018-03-05
### Changed
- Minor fixes on the check token method. Now the method sets the auth storage

## [1.1.19] - 2018-03-02
### Changed
- Minor fixes on the check token method and the http call

## [1.1.12/1.1.18] - 2018-03-01
### Changed
- Added check token
- Firebase import no more on notificationWeb
- All the Notification.getToken() functions makes a different callback sending the current object: notification = {'platform': token}
- HttpErrorHandler deals with text promises
- Error.mount now return promises

## [1.1.10/1.1.11] - 2018-02-26
### Changed
- Notification Web now receives a firebase instance pre-loaded

## [1.1.8/1.1.9] - 2018-02-26
### Changed
- AppBoostrap now sets the config.notification on the Notification class

## [1.1.4/1.1.7] - 2018-02-26
### Added
- Client notification for cordova and webapp

## [1.1.0/1.1.3] - 2018-02-16
### Added
- Error handler in Http: now the fetch errors are handled correctly
### Changed
- AppBootstrap now use the new Http.setup and set the callbacks for Http errors
- Deprecated warning on Http.setHeaders

## [1.0.25] - 2018-02-15
### Changed
- Http.reset and Http.setup created to simplify the bootstrap

## [1.0.24] - 2018-01-22
### Changed
- Text.formatNumber and Text.formatReal added

## [1.0.23] - 2018-01-10
### Changed
- VersionCheck.onNewVersion is now async

## [1.0.22] - 2018-01-09
### Changed
- storage.hasItem now reject on catch of getItem

## [1.0.21] - 2017-12-20
### Changed
- Created http download method

## [1.0.20] - 2017-12-19
### Changed
- Fixed http checkStatus and checkListener order

## [1.0.18/1.0.19] - 2017-12-12
### Changed
- Fixing relative data result

## [1.0.15/1.0.17] - 2017-11-27
### Changed
- Fixing rounding on DateTime.secondsToHour

## [1.0.14] - 2017-11-23
### Added
- relative and humanize added to DateTime

## [1.0.13] - 2017-11-17
### Changed
- AppBoostrap changed to work with mobile

## [1.0.12] - 2017-11-17
### Added
- AppBoostrap.getConfig()

## [1.0.11] - 2017-11-14
### Changed
- AppBoostrap build time on console

## [1.0.9] / [1.0.10] - 2017-11-14
### Added
- Object Handler for javascript objects methods

### Changed
- Fixes in export

## [1.0.8] - 2017-11-9
### Changed
- AppBootstrap() changed to AppBootstrap.setup()

## [1.0.3] / [1.0.7] - 2017-11-8
### Changed
- AppBootstrap fixes
- Http changed fetch to window.fetch.bind(window) - observe it

## [1.0.2] - 2017-11-7
### Added
- AppBootstrap included (src and test)

## [1.0.1] - 2017-10-31
### Added
- Some added

### Changed
- Some change

### Removed
- Some remove


src/http.js
import HttpErrorHandler from './httpErrorHandler';

class Http {
    static setup (version, client, env, contentType) {
        let headers = {};
        headers['Content-Type'] = contentType;
        headers['Accept'] = contentType;
        headers['client'] = client;
        headers['client-env'] = env;
        headers[client + '-version'] = version;
        Http.defaultHeaders = headers;
        Http.headers = headers;
    }

    static reset () {
        Http.headers = Http.defaultHeaders;
        Http.authorization = '';
        Http.params = {};
    }

    static setAuthorization (token) {
        Http.authorization = token;
        Http.headers = {...Http.defaultHeaders, authorization: token}
    }

    static setHeaderParam (key, value) {
        let headers = Http.headers;
        headers[key] = value;
        Http.headers = headers;
    }

    static setFetch (fetchMethod) {
        Http.fetch = fetchMethod;
    }

    static setBaseURL (baseURLSetup) {
        Http.baseURL = baseURLSetup;
    }

    static setHeaders (headersSetup) {
        console.warn('Http.setHeaders is deprecated. Consider use Http.setup(version, client, env, contentType) instead.');
        Http.headers = headersSetup;
    }

    static setEndpointParam (key, value) {
        Http.params[key] = value;
    }

    static getHeaders () {
        return Http.headers;
    }

    static getUrl (uri) {
        if (!Http.fetch) {
            throw new Error('This environment don\'t have the window.fetch. ' +
                'Please set a custom fetch method using Http.setFetch().');
        }
        let url = null;
        if (!uri.startsWith('http'))
            url = Http.baseURL + uri;
        else
            url = uri;
        const regex = /{.*}/i;
        let matches = url.match(regex);
        if (matches) {
            matches.map((match) => {
                let value = Http.params[match];
                url = url.replace(match, value);
            });
        }
        return url;
    }

    static post (uri, body) {
        let url = Http.getUrl(uri);
        return new Promise((resolve, reject) => {
                console.log(url + ' - POST - body ', body);
                // console.log(headers);
                body = JSON.stringify(body);
                Http.fetch(url, {
                    method: 'POST',
                    credentials: 'same-origin',
                    headers: Http.headers,
                    body: body
                }).then(Http.checkStatus)
                    .then(Http.checkListener)
                    .then(Http.parseJSON)
                    .then(response => {
                        console.log('POST > ' + uri + ' > response', response);
                        resolve(response);
                    }).catch(error => {
                    error = Http.errorHandler.handle(error);
                    reject(error);
                });
            }
        );
    }

    static put (uri, body) {
        let url = Http.getUrl(uri);
        return new Promise((resolve, reject) => {
                console.log(url + ' - PUT - body ', body);
                // console.log(headers);
                body = JSON.stringify(body);
                Http.fetch(url, {
                    method: 'PUT',
                    credentials: 'same-origin',
                    headers: Http.headers,
                    body: body
                }).then(Http.checkStatus)
                    .then(Http.checkListener)
                    .then(Http.parseJSON)
                    .then(response => {
                        console.log('PUT > ' + uri + ' > response', response);
                        resolve(response);
                    }).catch(error => {
                    error = Http.errorHandler.handle(error);
                    reject(error);
                });
            }
        );
    }

    static get (uri) {
        let url = Http.getUrl(uri);
        return new Promise((resolve, reject) => {
                console.log(url + ' - GET');
                Http.fetch(url, {
                    method: 'GET',
                    credentials: 'same-origin',
                    headers: Http.headers
                }).then(Http.checkStatus)
                    .then(Http.checkListener)
                    .then(Http.parseJSON)
                    .then(response => {
                        console.log(url + ' response', response);
                        resolve(response);
                    }).catch(error => {
                    error = Http.errorHandler.handle(error);
                    reject(error);
                });
            }
        );
    }

    static download (uri) {
        let url = Http.getUrl(uri);
        return new Promise((resolve, reject) => {
                console.log(url + ' - DOWNLOAD');
                Http.fetch(url, {
                    method: 'GET',
                    credentials: 'same-origin',
                    headers: Http.headers
                }).then(Http.checkListener)
                    .then(Http.checkStatus)
                    .then(response => {
                        resolve(response);
                    }).catch(error => {
                    error = Http.errorHandler.handle(error);
                    reject(error);
                });
            }
        );
    }

    static delete (uri) {
        let url = Http.getUrl(uri);
        return new Promise((resolve, reject) => {
                console.log(url + ' - DELETE');
                // console.log(headers);
                Http.fetch(url, {
                    method: 'DELETE',
                    credentials: 'same-origin',
                    headers: Http.headers
                }).then(Http.checkStatus)
                    .then(Http.checkListener)
                    .then(response => {
                        console.log(url + ' response', response);
                        resolve(response);
                    }).catch(error => {
                    error = Http.errorHandler.handle(error);
                    reject(error);
                });
            }
        );
    }

    static checkStatus (response) {
        // console.log('checkStatus');
        if (response.status >= 200 && response.status <= 210) { // 200 - OK || 201 - Created
            return response;
        } else {
            return Http.errorHandler.mount(response);
        }
    }

    static checkListener (response) {
        if (Http.requestListener) {
            Http.requestListener(response);
        }
        return response;
    }

    static parseJSON (response) {
        // console.log('parseJSON');
        return response.json();
    }

    static setRequestListener (callback) {
        Http.requestListener = callback;
    }

    static setErrorHandler (ErrorHandlerClass) {
        Http.errorHandler = ErrorHandlerClass;
    }
}

Http.headers = {};
Http.defaultHeaders = {};
Http.authorization = '';
Http.baseURL = '';
Http.params = {};
Http.requestListener = null;
Http.errorHandler = HttpErrorHandler;
Http.fetch = typeof fetch !== 'undefined' ? window.fetch.bind(window) : null;

export default Http;
